src/resolver.ts

```diff
@@ -11,35 +16,60 @@ async function get(url: string): Promise<any> {
   return res.json()
 }
 
-export function getResolver() {
+export function getResolver(): Record<string, DIDResolver> {
   async function resolve(
     did: string,
     parsed: ParsedDID
-  ): Promise<DIDDocument | null> {
-    let path = parsed.id + DOC_PATH
+  ): Promise<DIDResolutionResult> {
+    let err = null
+    let path = decodeURIComponent(parsed.id) + DOC_PATH
     const id = parsed.id.split(':')
-    if (id.length > 1) path = id.join('/') + '/did.json'
+    if (id.length > 1) {
+      path = id.map(decodeURIComponent).join('/') + '/did.json'
+    }
+
     const url: string = `https://${path}`
 
-    let data: any = null
-    try {
-      data = await get(url)
-    } catch (error) {
-      throw new Error(
-        `DID must resolve to a valid https URL containing a JSON document: ${error.message}`
-      )
-    }
+    const didDocumentMetadata = {}
+    let didDocument: DIDDocument | null = null
 
-    const docIdMatchesDid = data.id === did
-    if (!docIdMatchesDid) {
-      throw new Error('DID document id does not match requested did')
-    }
+    do {
+      try {
+        didDocument = await get(url)
+      } catch (error) {
+        err = `DID must resolve to a valid https URL containing a JSON document: ${error}`
+        break
+      }
+
+      // TODO: this excludes the use of query params
+      const docIdMatchesDid = didDocument?.id === did
+      if (!docIdMatchesDid) {
+        err = 'DID document id does not match requested did'
+        break
+      }
+    } while (false)
```

Comment:
do {} while loop seems quite non-standard. while(false) never executes that block a second time. What's the reasoning here?

Reply:
This is a technique useful for sequential checks where it would not make sense to continue eny code after a failure.
This is achieved using the `break` statements to jump out of the block in case it no longer makes sense to continue.

This can also be done by wrapping things in `try/catch` and using `throw` for control flow, but this introduces overhead. And it can also be done using lots of `if` checks which adds a lot of code.

Reply:
Hm, I understand. Not sure if I like that pattern too much though. If there were a lot of break statements in there than maybe I'd understand the usage here more, but in general that's only on break in here. (that last one can be removed, correct?)

Reply:
technically correct. I kept it to protect from future checks.